### PR TITLE
[Feature Request] 9-patch drawing function

### PR DESCRIPTION
**Pull request by overdev**
# Feature Request: 9-patch drawing function

This is a great feature that is always very useful for games, specially when
it comes to GUI rendering. I took some time thinking on how it would fit in
raylib, so I came up with the piece of code below that may or may not serve as
pull request. Let me give you a brief description first.

## New structure: `NinePatch`

These are not required, but could make things easier. Their function is to
hold the (minimum) necessary information about how a n-patch should be
drawn without having to pass a lot of the same arguments to the call
all the time:

```c
typedef struct NinePatch {
    Texture2D texture;      // The texture associated with the 9-patch (maybe Texture2D *, instead?)
    Rectangle sourceRec;    // The 9-patch region in the texture
    Vector2 minSize;        // The minimum size the 9-patch can be shrunk to
    float borderWidth[4];   // The widths of the left, top, right and bottom borders
    int padding[4];         // Helps the n-patch contents fit nicely inside
} NinePatch;
```

## New function: `DrawNinePatch`

Based on the structures defined above, the n-patch drawing function would
assume the following signature:
```c
void DrawNinePatch(NinePatch ninePatch, Rectangle destRec, bool usePadding, Vector2 origin, float rotation, Color tint) {
    // Draws the given ninePatch at the position and size specified by destRec, tinted with given color,
    // optionally rotated at rotation degrees around origin vector.
    // If the destRec width and/or height is smaller than NinePatch.minSize x and/or y values,
    // minSize value(s) will be used.
    // This version ignores the padding
}
```

## The implementation: a hack of `DrawTexturePro`

`DrawTexturePro` is the function that gave the insights on how this
features could be implemented. I might be mistaken but with the
necessary tweaks and hacks, as demonstrated below, `DrawTexturePro` can
bring `DrawNinePatch` to existence.

### Tweak 1: `sourceRec` and `texture`

The first basic changes is to `texture` and `sourceRec` wich now refers to the
`ninePatch`'s struct `texture` and `sourceRec`:
```c
// ...
// Check if n-patch texture is valid
if (ninePatch.texture.id > 0)
{
    float width = (float)ninePatch.texture.width;
    float height = (float)ninePatch.texture.height;
    
    if (ninePatch.sourceRec.width < 0) ninePatch.sourceRec.x -= ninePatch.sourceRec.width;
    if (ninePatch.sourceRec.height < 0) ninePatch.sourceRec.y -= ninePatch.sourceRec.height;

    rlEnableTexture(ninePatch.texture.id);
    // ...
```

### Tweak 2: Adjusting TexCoords and Vertices

I was wondering if wouldn't be better to set the draw mode to `RL_TRIANGLE_STRIP` but, as we can
see in `rlgl.h`, that primitive type (or draw mode) is not defined. It could save some calls to
`rlTexCoord2f` and rlVertex2f (maybe 12 of each on `DrawNinePatch` and 4 of each in `DrawThreePatch`).
That's not a problem, anyways.

Prior to calculating the coordinates, we have to understand what to do when destRec size is set to a
value that's too small.

#### Shrinking down to (0, 0) size

As we know, the corners of a n-patch does not change in side, the sides may change in one or other
axis and the center part always changes in size in both axis. It's not very clear, though, how the n-patch
corners should behave when the `destRec` width and/or height is smaller than left + right
and/or top + bottom `borderWidth`s.

The approach in this case was to crop the corners' and sides' borderWidth in proportion to its
values. So, for a 32x32 source rect size with 8 pixels for the width of left border and 16 for the right,
when the n-patch destRec width is less than 32 pixels, 15 pixels let's say, the left border is
adjusted to 5 pixels and the right border is adjusted to 10 pixels. The respective code is this:

```c
// ...
bool drawCenter = true;
bool drawMiddle = true;

float borderWidth[4];  // copy the ninePatch.borderWidth[4] values so they can be adjusted
for (int i = 0; i < 4; i++) { borderWidth[i] = ninePatch.borderWidth[i]; }

// adjust the lateral (left and right) border widths if destRec.width < ninePatch.texture.width
if (destRec.width <= ninePatch.borderWidth[0] + ninePatch.borderWidth[2])
{
    drawCenter = false;
    borderWidth[0] = (borderWidth[0] / (ninePatch.borderWidth[0] + ninePatch.borderWidth[2])) * (borderWidth[0] + borderWidth[2]);
    borderWidth[2] = destRec.width - borderWidth[0];
}
// adjust the lateral (top and bottom) border heights if destRec.height < ninePatch.texture.height
if (destRec.height <= ninePatch.borderWidth[1] + ninePatch.borderWidth[3])
{
    drawMiddle = false;
    borderWidth[1] = (borderWidth[1] / (ninePatch.borderWidth[1] + ninePatch.borderWidth[3])) * (borderWidth[1] + borderWidth[3]);
    borderWidth[3] = destRec.height - borderWidth[1];
}
// ...
```
This piece of code goes right after the piece in Tweak 1. After the necessary adjustements,
the vertex and texture coords are calculated:

```c
Vector2 vertA, vertB, vertC, vertD;
vertA.x = 0.0f;
vertA.y = 0.0f;
vertB.x = borderWidth[0];
vertB.y = borderWidth[1];
vertC.x = destRec.width  - borderWidth[2];
vertC.y = destRec.height - borderWidth[3];
vertD.x = destRec.width;
vertD.y = destRec.height;

Vector2 coordA, coordB, coordC, coordD;
coordA.x = sourceRec.x / width;
coordA.y = sourceRec.y / height;
coordB.x = (sourceRec.x + ninePatch.sourceRec.borderWidth[0]) / width;
coordB.y = (sourceRec.y + ninePatch.sourceRec.borderWidth[1]) / height;
coordC.x = (sourceRec.x + ninePatch.sourceRec.width  - borderWidth[2]) / width;
coordC.y = (sourceRec.y + ninePatch.sourceRec.height - borderWidth[3]) / height;
coordD.x = (sourceRec.x + ninePatch.sourceRec.width)  / width;
coordD.y = (sourceRec.y + ninePatch.sourceRec.height) / height;
```

That's a lot of vectors, maybe... There's actually two pairs of AABBs, the first
formed by A and D vectors (of vertices and coords) and, the second, formed by B and C
vectors. From these AABBs all the n-patch quads are made. In the last tweak, shown
below, we see how it's straightforward to pass the coordinates.

### Last Tweak: sending the vertices and texCoords

Depending on the size of sourceRec, 4, 6 or 9 quads are draw. The drawing is done
from left to right, top to bottom. The piece below shows the code of the top
quads (top-left, top-center and top-right):

```c
// ------------------------------------------------------------
// TOP-LEFT QUAD
// Bottom-left corner for texture and quad
rlTexCoord2f(coordA.x, coordB.y); rlVertex2f(vertA.x, vertB.y);

// Bottom-right corner for texture and quad
rlTexCoord2f(coordB.x, coordB.y); rlVertex2f(vertB.x, vertB.y);

// Top-right corner for texture and quad
rlTexCoord2f(coordB.x, coordA.y); rlVertex2f(vertB.x, vertA.y);

// Top-left corner for texture and quad
rlTexCoord2f(coordA.x, coordA.y); rlVertex2f(vertA.x, vertA.y);

if (drawCenter)
{
    // TOP-CENTER QUAD
    // Bottom-left corner for texture and quad
    rlTexCoord2f(coordB.x, coordB.y); rlVertex2f(vertB.x, vertB.y);

    // Bottom-right corner for texture and quad
    rlTexCoord2f(coordC.x, coordB.y); rlVertex2f(vertC.x, vertB.y);

    // Top-right corner for texture and quad
    rlTexCoord2f(coordC.x, coordA.y); rlVertex2f(vertC.x, vertA.y);

    // Top-left corner for texture and quad
    rlTexCoord2f(coordB.x, coordA.y); rlVertex2f(vertB.x, vertA.y);
}

// TOP-RIGHT QUAD
// Bottom-left corner for texture and quad
rlTexCoord2f(coordC.x, coordB.y); rlVertex2f(vertC.x, vertB.y);

// Bottom-right corner for texture and quad
rlTexCoord2f(coordD.x, coordB.y); rlVertex2f(vertD.x, vertB.y);

// Top-right corner for texture and quad
rlTexCoord2f(coordD.x, coordA.y); rlVertex2f(vertD.x, vertA.y);

// Top-left corner for texture and quad
rlTexCoord2f(coordC.x, coordA.y); rlVertex2f(vertC.x, vertA.y);
```

> NOTE: the code examples above might be a little different from the sources.

That's it. Here's some very basic screenshots of the results:

image

image

Unfortunately the 9-patches `destRec` size can't be smaller than 12x12 without distorting. Also, I'm not sure this is the best algorithm -- I'm grateful it's working because I'm not a C programmer.

Lastly, the preview images above are from a custom build and I run it from an example written in my [raylibpy](https://github.com/overdev/raylibpy) Python binding wich will be released very soon.

### CODE REVIEW

**Comment by overdev** [comment]
Good news, it seems.

**Comment by raysan5** [comment]
@overdev Congrats! Trully an amazing work! It was really a long time since I saw such a well developed, detailed and organized implementation and explanation. It's a great addition and I want to review it carefully but I'll be adding it.

Please, could you also add the example and screenshots to [examples/textures](https://github.com/raysan5/raylib/tree/master/examples/textures).

Really, I'm very impressed for your care on the implementation, please, keep me updated if you thing of new features or idea. raylibpy also looks great! :)

**Comment by overdev** [comment]
Thanks @raysan5. This is the first pull request I ever made.

Yes, please, take a look at the code as soon as you can. I need the supervision of someone with experience in the language. Also, I hope it does not break any design rule so, please, make the changes you feel necessary — this PR is still just a suggestion.

Anyways, I will elaborate a better example and resouce image because I wrote that one in a hurry.

**Comment by overdev** [comment]
Oh, I forgot to mention. The initial idea was to add 9-patch and 3-patch. 3-patches can be resized only vertically or horizontally. That's why I used the term n-patch sometimes and some words in plural.

Maybe, extending this feature to include it would require a small tweak in NinePatch struct, another in the drawing function and the addiction of a new enum:
```c
typedef struct NPatchType
{
    NPT_9PATCH = 0,  // defaults to 9-patch
    NPT_3PATCH_VERTICAL,
    NPT_3PATCH_HORIZONTAL
}

typedef struct NPatch   // renamed
{
    // Here goes the fields we already defined
    NPatchType type;
} NPatch;

// then
RLAPI void DrawNPatch(...);
```

I don't know the PR system well enough. If I do these changes, will they be visible in this PR?

**Comment by OvermindDL1** [comment]
> I don't know the PR system well enough. If I do these changes, will they be visible in this PR?

As long as you update the same branch, yes.

**Comment by overdev** [comment]
Surprise, surprise! I'm having a hard time translating the examples to C...

**Comment by raysan5** [comment]
Great work @overdev! Just waiting for the example to integrate...

Some details:
```c
typedef struct NPatch {
    Texture2D texture;      // The texture associated with the 9-patch (maybe Texture2D *, instead?)
    Rectangle sourceRec;    // The 9-patch region in the texture
    Vector2 minSize;        // The minimum size the 9-patch can be shrunk to
    float borderWidth[4];   // The widths of the left, top, right and bottom borders
    int padding[4];         // Helps the n-patch contents fit nicely inside
    int type;               // The type of this n-patch: 9-patch, 3-patch vertical or 3-patch horizontal
} NPatch;
```
I don't think `minSize` should be a NPatch parameter, I consider it more an implementation detail, internal to raylib function, it could be just a define...

`float borderWidth[4]`, why not an `int`, calling it Width could be a bit confusing for users, maybe call it `borderSize`?

`padding[4]` seems a bit confusing for me, if you draw to a destRectangle, `borderSize` data should be enough... maybe I didn't understand it correctly...

Also thinking about moving `Texture2D` and `sourceRec` outside the struct for consistency with other `DrawTexture()` function... what do you think?
```c
RLAPI void DrawTextureNPatch(Texture2D texture, Rectangle srcRec, Rectangle destRec, NPatch nPatch, Vector2 origin, float rotation, Color tint);
```

**Comment by overdev** [comment]
In the good old times of Window XP era, with all that fuzz about Windows themes and what not, I was one of those people obsessed to understand how the theme files worked, what it contained et cetera. With the help of some theme editing tools I learned for the first time what was a 9-patch. I was young and fool (more than I am today, people say) and lost hours just of my life just starring at this huge INI file that (along with all the resource images) compose the *.uxtheme. "That is so awesome!" I used to think.

TL; DR, I think those *.uxtheme files with all its keys and whatever got stuck in my head. I think my imagination flew too high and the idea of a simplistic theme mechanism is floating in my head.

The function of NPatch the way it is now is somewhat similar to the CharInfo in the sense that a CharInfo is to a font what a NPatch is to a large UI texture, like this GWEN UI skin:

image

You'll want to define the NPatchs for each element only once and then make different skins textures following the format specified. All fields I added to NPatch I did having this in mind.

Now, I don't think it has to be this way. I would prefer a more lean struct and, letting this UI skin stuff aside, I have no doubt it can be achieved. Let me talk a litle about the meaning of each field and why it could stay or leave:
* `Texture2D texture;` - In fact, it has no strong reason to be. That is because, in a drawing function, it doesn't matter too much. Also, one may want to change the texture to draw using the same NPatch. It can go easily. (Only in that boring uxtheme subject it has some importance.)
* `Rectangle sourceRec;` - I opted to add this field because it is less likely to change in a good number of cases, in contrast to `destRec`, wich changes almost every time. I could go too but its presence can help a lot.
* `Vector2 minSize;` - It could go too. Its presence serves only as a general rule, I mean, only when you'll never want to allow the `destRec` size to go below certain values.
* `float borderWidth[4];` I got the same feelings about the name. It is quite confusing, so it's better to change its name. (English is not my natural language, so I struggle to find the right words most of the time.) I chose the float type because it is used as a texture coordinate. This is specially important when the `sourceRec` width and/or height goes below the `sourceRec`'s; because in that case the `borderWidth[x]` is not always integer (due to the n-patch shrink method). I would suggest a NPatch struct where only `sourceRec` and `borderWidth` fields are defined as these two are ones that really matters (and split `borderWidth[4]` into individual fields, maybe). By the way, `borderWidth[]` defines the 4 or 2 lines that cut `sourceRec` into 9 or 3 quads (like a tic-tac-toe grid). These lines are relative to their sides, e.g., the `borderWidth[2]` (a vertical cut at the right) is an offset of the `sourceRec`'s right, so `borderWidth[2] == 0 == (sourceRec.x + sourceRec.width)` must be true. Nonetheless, you have far more knowledge in C than I have. You may have good reasons on why it should be `int`.
* `int padding[4];` - In those cases where you want `destRec` to represent the bounds of some content, instead some large region where this content is placed. For example, when rendering a button like thing that auto resizes to fit its caption, the padding is that spacing between the text bounds and the button bounds. By setting a padding of (6, 4, 6, 4), the resulting 9-patch will be 12 pixels larger in width and 8 in height. It is useful in some few cases, so it can go without too much thinking.
* `int type;` - This could also go away if we split DrawNPatch into two dedicated functions. The orientation (whether the 3-patch layout is vertical or horizontal) could then be turned into a parameter.

One more thing: `DrawNPatch` should also be renamed, maybe to `DrawTextureNPatch`, to follow the naming convention. I have to apologize. I did everything in a hurry, should have thought more carefully.

Finally, one proposal:
```c
typedef struct NPatch {
    Rectangle sourceRec;  // the n-patch region in the texture
    int top;       // vertical cut offset from rec top to bottom
    int left;         // horizontal cut offset from rec left to right
    int right;       // vertical cut offset from rec right ro left
    int bottom;     // horizontal cut offset from rec bottom to top
}

// also:
void DrawTexture9Patch(Texture2D texture, NPatch ninePatch, Rectangle destRec, Vector2 origin, float rotation, Color tint);

void DrawTexture3Patch(Texture2D texture, NPatch ninePatch, bool IsVertical, Rectangle destRec, Vector2 origin, float rotation, Color tint);

// maybe (without using NPatch struct):
void DrawTexture9PatchEx(Texture2D texture, Rectangle sourceRec, int left, int top, int right, int bottom, Rectangle destRec, Vector2 origin, float rotation, Color tint);

void DrawTexture3PatchEx(Texture2D texture, Rectangle sourceRec, int start, int end, bool isVertical, Rectangle destRec, Vector2 origin, float rotation, Color tint);
```

Whatever you choose is OK to me.

**Comment by raysan5** [comment]
You give me an idea... Maybe we can use ` struct NPatchInfo` similar to `CharInfo`, I mean:
```c
typedef struct NPatchInfo {
    Rectangle sourceRec;  // the n-patch region in the texture
    int top;              // vertical cut offset from rec top to bottom
    int left;             // horizontal cut offset from rec left to right
    int right;            // vertical cut offset from rec right ro left
    int bottom;           // horizontal cut offset from rec bottom to top
    int type;        // n-patch type
} NPatchInfo;
```
And again just simplify function to:
```c
RLAPI void DrawTextureNPatch(Texture2D texture, NPatch nPatch, Rectangle destRec, Vector2 origin, float rotation, Color tint);
```
What do you think?

PS. If you like GUI systems, probably you'll like to take a look to [raygui](https://github.com/raysan5/raygui) (raylib-based IMGUI system).

**Comment by overdev** [comment]
Yeah, I'm seriously thinking in making a raygui python binding for it or add it as a raylibpy extra module (wich would be easier for me to do). In fact, this whole 9-patch think came to mind after I took a good look in the raygui repo. Awesome work you got there.

I'll add your suggestion. The code is mostly finished. I divided the function into `DrawTexture9Patch` and DrawTexture3Patch (with `Ex` versions as mentioned above) so `int type;` is now a `bool isVertical` param in `DrawTexture3Patch[Ex]`.

Everything compiled with no problems so far (the compiler doesn't like the things I write to it :sweat_smile: ) and I'm rewriting the examples to reflect the changes. As soon as I finish I'll commit the changes.

**Comment by raysan5** [comment]
Hi @overdev, I really prefer to limit it to just one function `DrawTextureNPatch()`, if we're using the `NPatchInfo` struct, we can just add required n-patch info there and leave the function simple. Users are more prone to remember just one function.

**Comment by overdev** [comment]
Oh, that's not a problem! I'll do it.

**Comment by overdev** [comment]
I believe this PR is now very close to the library standards. Hopefully, any further corrections will be minimal.

**Comment by raysan5** [comment]
Amazing work @overdev! Maybe I review some formatting or naming but it's probably one of the most detailed and cared PR I've received! Thank you very much! :D

**Comment by overdev** [comment]
It's me who have to say thank you, @raysan5! Thanks for the patience, above all else! I say this because I'm completely new to this community development thing and I have to be careful on how I approach other people's projects with my ideas. I have a lot to learn.

I have some other PR ideas but I'll have to make some research first (to see if I'm up to the task, mostly). In parallel to this, I'm planning to add _rayGui_ as a _raylibpy_ submodule. If everything goes fine, soon we'll be talking again. :+1:
